onnxruntime/python/tools/transformers/convert_generation.py

```diff
@@ -1066,6 +1075,35 @@
     return subg
 
 
+def update_decoder_subgraph_use_decoder_masked_multihead_attention(subg):
+    """Update the Attention nodes to DecoderMaskedMultiheadAttention.
+
+    Args:
+        subg (GraphProto): GraphProto of the decoder subgraph
+    """
+    decoder_masked_attention_supported_attr = [
+        "past_present_share_buffer",
+        "num_heads",
+        "qkv_hidden_sizes",
+        "scale",
+        "domain",
+    ]
+    new_nodes = []
+    for node in subg.node:
+        if node.op_type == "Attention":
+            kwargs = kwargs_of(node)
+            for k in kwargs.copy():
+                if k not in decoder_masked_attention_supported_attr:
+                    del kwargs[k]
+            nis = []
+            nis.extend(node.input)
+            node = onnx.helper.make_node("DecoderMaskedMultiheadAttention", nis, node.output, name=node.name, **kwargs)
```

Comment:
We have several TODOs (like head_size !=v_head_size) in the implementation. If we don't support them now, we should not switch to DecoderMaskedMultiheadAttention here if qkv_hidden_sizes exists and qk and v have different head size.

Reply:
That's true (and I can add such gating logic) but it is hard to add all gating logic in the conversion tool. For example, the kernel also doesn't support anything other than 2-D mask and doesn't support relative positional bias as well (other TODOs). But, we won't know the kind of mask the user is using and if they are using rel postiional bias until runtime (when the switch to DecoderMaskedMultiheadAttention is already done).

Reply:
Added gating logic in the script around `qkv_hidden_sizes `

Reply:
You might notice that qkv_hidden_size attribute is not added by the fusion [fusion_gpt_attention.py](https://github.com/microsoft/onnxruntime/blob/f88b97ede272e54d67e7f4da165359df4b534a4a/onnxruntime/python/tools/transformers/fusion_gpt_attention.py#L199-L207). So, we can just skip here if the attribute qkv_hidden_size exists.

Reply:
I see, thanks for the info. So, which component adds qkv_hidden_size if it all it has to be added ?